lib/multipass_ex/crypt.ex

```diff
@@ -0,0 +1,116 @@
+defmodule MultipassEx.Crypt do
+  @moduledoc """
+  Module for encrypting and decrypting data in compliance with the Ruby MultiPass
+  gem.
+
+  AES_CBC128 is the encryption algorithm used.
+
+  __Warning__: The ruby gem uses an insecure IV and it's mirrored here to be compatible
+  since it is not sent with the Multipass data.
+  """
+  require Logger
+
+  alias MultipassEx.{DecodingError, EncodingError, CryptError}
+  alias MultipassEx.Coding
+
+  @algorithm :aes_cbc128
+  @block_size 16
+  @key_byte_length 16
+  @iv String.duplicate(<<0>>, 16)
+
+  @spec encrypt(map(), binary(), binary()) ::
+          {:ok, binary()}
+          | {:error, EncodingError.t()}
+  def encrypt(data, site_key, api_key) do
+    key = generate_key(site_key, api_key)
+
+    prepared_data = prepare_data_for_encryption(data)
+
+    encrypted_data =
+      @algorithm
+      |> :crypto.block_encrypt(key, @iv, prepared_data)
+      |> Coding.encode()
+
+    {:ok, encrypted_data}
+  rescue
+    Jason.EncodeError ->
+      {:error,
+       %EncodingError{
+         message: "json encoding failed",
+         reason: :json_encoding_error
+       }}
+  end
+
+  @spec decrypt(binary(), binary(), binary()) ::
+          {:ok, map()} | {:error, CryptError.t() | DecodingError.t()}
+  def decrypt(data, site_key, api_key) do
+    case Coding.decode(data) do
+      {:error, %DecodingError{} = error} ->
+        {:error, error}
+
+      {:ok, decoded_data} ->
+        decrypt_or_error(decoded_data, site_key, api_key)
+    end
+  end
+
+  @spec decrypt_or_error(binary(), binary(), binary()) ::
+          {:ok, map()} | {:error, CryptError.t() | DecodingError.t()}
+  defp decrypt_or_error(data, site_key, api_key) do
+    key = generate_key(site_key, api_key)
+
+    processed_data =
+      @algorithm
+      |> :crypto.block_decrypt(key, @iv, data)
+      |> process_decrypted_data()
+
+    {:ok, processed_data}
+  rescue
+    ArgumentError ->
```

Comment:
as from above, i think you should allow the encryption function determine if it should return an error